cluster-agent/controllers/managed-gitops/eventloop/operation_event_loop_test.go

```diff
@@ -1298,13 +1342,96 @@ var _ = Describe("Operation Controller", func() {
 						return fmt.Errorf(expectedErr)
 					},
+					refreshApp: refreshApplication,
+				}
+
+				retry, err := task.PerformTask(ctx)
+				Expect(err.Error()).Should(Equal(expectedErr))
+				Expect(retry).To(BeTrue())
+			})
+
+			It("should return an error and retry if the refresh fails", func() {
+				By("create a SyncOperation in the database")
+				syncOperation := db.SyncOperation{
+					SyncOperation_id:    "test-syncoperation",
+					Application_id:      applicationDB.Application_id,
+					DeploymentNameField: "test",
+					Revision:            "main",
+					DesiredState:        db.SyncOperation_DesiredState_Running,
+				}
+				err = dbQueries.CreateSyncOperation(ctx, &syncOperation)
+				Expect(err).To(BeNil())
+
+				By("create Operation DB row and CR for the SyncOperation")
+				createOperationDBAndCR(syncOperation.SyncOperation_id, gitopsEngineInstanceID)
+
+				By("check if the sync failed error is returned with retry")
+				expectedErr := "failed to refresh app: operation"
+				task.syncFuncs = &syncFuncs{
+					refreshApp: func(ctx context.Context, c client.Client, s1, s2 string) error {
+						return fmt.Errorf("failed to refresh app: %s", s1)
+					},
 				}
 
 				retry, err := task.PerformTask(ctx)
 				Expect(err.Error()).Should(Equal(expectedErr))
 				Expect(retry).To(BeTrue())
 			})
 
+			It("should handle conflicts while refreshing application", func() {
+				By("create a SyncOperation in the database")
+				syncOperation := db.SyncOperation{
+					SyncOperation_id:    "test-syncoperation",
+					Application_id:      applicationDB.Application_id,
+					DeploymentNameField: "test",
+					Revision:            "main",
+					DesiredState:        db.SyncOperation_DesiredState_Running,
+				}
+				err = dbQueries.CreateSyncOperation(ctx, &syncOperation)
+				Expect(err).To(BeNil())
+
+				By("create Operation DB row and CR for the SyncOperation")
+				createOperationDBAndCR(syncOperation.SyncOperation_id, gitopsEngineInstanceID)
+
+				By("introduce a conflict")
+				appCRClone := applicationCR.DeepCopy()
+				appCRClone.Annotations = make(map[string]string)
+				appCRClone.Annotations = map[string]string{
+					"key-1": "value-1",
+					"key-2": "value-2",
+				}
+				err = k8sClient.Update(ctx, appCRClone)
+				Expect(err).To(BeNil())
+
+				applicationCR.Annotations = make(map[string]string)
+				applicationCR.Annotations["key-3"] = "value-3"
+				err = k8sClient.Update(ctx, applicationCR)
+				Expect(apierr.IsConflict(err)).To(BeTrue())
+
+				task.syncFuncs = &syncFuncs{
+					appSync: func(ctx context.Context, s1, s2, s3 string, c client.Client, cs *utils.CredentialService, b bool) error {
+						return nil
+					},
+					refreshApp: refreshApplication,
+				}
+				retry, err := task.PerformTask(ctx)
+				Expect(err).Should(BeNil())
+				Expect(retry).To(BeFalse())
+
+				By("verify if the refresh annotation is removed")
```

Comment:
It seems like this `Eventually` is testing the goroutine the refreshHandler goroutine that is started in `BeforeEach`, rather than testing the product code itself.

I would think we would want to instead test if the refresh annotation was added.